ARM64-SVE: Add checks for MOVPRFX sequencing (#105514)

a74nh opened the pull request:
As per the Arm Arm, `MOVPRFX` instructions must be followed by an SVE instruction of specific type, otherwise behaviour is undefined.

Add a function which checks that two instructions in sequence are valid. For now this only cares when the first instruction is a `MOVPRFX`, but the function could be expanded to check for any pairs, should we need to.

Function is called after code generation in a loop which cycles through all `IG`s and all instructions.
This ensures the stream is valid after peephole type optimisations.

Alternatively, this function could be called from `emitInsSanityCheck()`. `EMIT_BACKWARDS_NAVIGATION` can then be used to get the previous instruction. However, defining this adds an extra variable to `instrDesc`, reducing the small immediate size from 7bits to 2bits.

Alternatively, this function could be inside every `emitIns_` function, allowing `emitLastIns` to be used for the previous instruction. However, that is a lot of call sites and some could easily be missed. In addition a peephole optimisation could still occur after this point.

Currently requires #106125 for all SVE tests to pass.
AddSequentialAcross tests disabled until #106180 is fixed.

Resolves: https://github.com/dotnet/runtime/issues/105514

a74nh committed:
ARM64-SVE: Add checks for MOVPRFX sequencing

a74nh commented:
@dotnet/arm64-contrib @kunalspathak

a74nh committed:
Fix formatting

kunalspathak commented on src/coreclr/jit/emit.cpp:
```diff
@@ -7095,8 +7095,23 @@ unsigned emitter::emitEndCodeGen(Compiler*         comp,
+
+#if defined(DEBUG) && defined(TARGET_ARM64)
+        instrDesc* currId = emitFirstInstrDesc(ig->igData);
+        for (unsigned cnt = ig->igInsCnt; cnt > 0; cnt--)
```
i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted *last*). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

a74nh replied:
> this will not catch case where `movprfx` was in previous IG and following SVE instruction is in new IG.

That should be ok. `prevId` is defined outside the IG loop and isn't reset ever. So on the next IG, `prevId` should be the last instruction in the previous `IG`.

> i think probably a better thing would be to track `emitPrevInstr` that we discussed offline. It will get set to `emitLastIns` just before it gets reset and is used in `emitInsSanityCheck()` ( `emitPrevInstr` will be `movprfx` and `emitLastIns` will be current instruction that was emitted _last_). As soon as the validation is done, reset `emitPrevInstr = nullptr` so we do not use it accidently.

I remember trying to do that when we did the peephole optimisations, and there was push back. I can't remember the issues around it. But maybe that is why `EMIT_BACKWARDS_NAVIGATION` was added instead?

kunalspathak replied:
> prevId is defined outside the IG loop and isn't reset ever

that should be fine then.

> that is why EMIT_BACKWARDS_NAVIGATION was added instead

It was added specifically to track more than 1 last instruction for xarch zero extension peephole IIRC. I realized that we also have `emitRemoveLastInstruction()` so perhaps doing the validation once everything is emitted makes sense.

kunalspathak reviewed (changes requested):
Suggested some changes. Also, can you also confirm all stress tests are passing?

a74nh committed:
Move emitInsPairSanityCheck

a74nh commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -957,6 +957,208 @@ void emitter::emitInsSanityCheck(instrDesc* id)
+
+        case IF_SVE_CC_2A: // <Zdn>.<T>, <V><m>
+        case IF_SVE_FU_2A: // <Zda>.<T>, <Zn>.<T>, #<const>
+            assert(secondId->idReg1() != secondId->idReg2());
```
> i was hoping to see lot more validation for registers of firstId

That should be all at the bottom of the function. I'll double check that I have everything, because the text you have above is a much more verbose version than what I was using.

kunalspathak replied:
looks like you have most of the things covered. We need to add validation for:

- MOVPRFX cannot be the last instruction in a sequence
- The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)
- MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

a74nh replied:
> * MOVPRFX cannot be the last instruction in a sequence

Done.

> * The predicated instruction at PC+4 must use the merging predicate. (not sure if this is taken care by the format checks)

Looking at all the formats, they are all <Pg>/M or unpredicated. I've added some extra checking to make sure they correspond to the correct movprfx instruction.

> * MOVPRFX destination register MUST be used as the destination register in the instruction at PC+4, "**and is not allowed to be used in any other position other than destructive input**"

That's done in the switch. Eg:
`            assert(secondId->idReg1() != secondId->idReg3());`
I had to do it there because you need the format to know what types the registers are.

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,207 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+            break;
+
+        default:
+            assert(false);
```
```suggestion
            assert(!"Got unexpected instruction format after MOVPRFX");
```

a74nh committed:
Improve asserts

a74nh committed:
check final instruction

a74nh committed:
Merge main

Change-Id: Ib78085c631ade5e923cddfda4def7efb362b4587

a74nh committed:
Reorder switch table and add movprefxIsPredicated check

a74nh committed:
fix formatting

kunalspathak commented on src/coreclr/jit/emit.cpp:
```diff
@@ -7095,8 +7095,28 @@ unsigned emitter::emitEndCodeGen(Compiler*         comp,
+        // Additional check for the final instruction
+        if (ig->igNext == nullptr)
+        {
+            emitInsPairSanityCheck(prevId, nullptr);
```
```suggestion
            assert(prevId->idIns() != INS_sve_movprfx);
```

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+    // The movprfx can't be the last instruction
+    if (secondId == nullptr)
+    {
+        assert(!"Last instruction generated is a MOVPRFX");
```
```suggestion
       
```

not needed any more if you just check it in the caller.

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+    }
+    else
+    {
+        // 2 operand version
```
```suggestion
        // unpredicated version
```

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+        assert(firstId->idInsFmt() == IF_SVE_BI_2A);
+    }
+
+    // Quoted sections are taken fron the Arm manual.
```
```suggestion
    // Quoted sections are taken from the Arm manual.
```

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+        assert(firstId->idInsFmt() == IF_SVE_BI_2A);
+
+        // "The prefixed instruction must specify the same destination vector as the MOVPRFX instruction."
+        assert(firstId->idReg1() == secondId->idReg1());
```
This assert can be outside if-else.

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+    else
+    {
+        // 2 operand version
+        assert(firstId->idInsFmt() == IF_SVE_BI_2A);
```
```suggestion
        
```

no need of this validation. we are already doing it above.

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+            break;
+    }
+
+    if (firstId->idInsFmt() == IF_SVE_AH_3A)
```
```suggestion
    if (movprefxIsPredicated)
```

kunalspathak commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18421,4 +18421,243 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+        assert(firstId->idReg2() == secondId->idReg2());
+
+        // "and have the same maximum element size (ignoring a fixed 64-bit "wide vector" operand),"
+        assert(firstId->idInsOpt() == secondId->idInsOpt());
```
can this be true for unpredicated version as well?

a74nh replied:
The unpredicated `movprfx` doesn't care about element sizes, it is just `MOVPRFX <Zd>, <Zn>`.

kunalspathak reviewed (dismissed):
Added some more comments on validation.

a74nh committed:
fixups

a74nh committed:
Remove GenerateHWIntrinsicTests_Arm.cs changes

a74nh committed:
move idReg1() check

a74nh commented:
Remove the changes to `GenerateHWIntrinsicTests_Arm.cs`.

This PR needs to wait for #106292 (and #106125) until it can be merged

a74nh committed:
Merge main

Change-Id: I0a421eacf4895cfb2a30aaa6c483f4a7a3f17dd1

a74nh commented:
> This PR needs to wait for #106292 (and #106125) until it can be merged

Those are merged now. There is nothing blocking this PR.

However - when running the stress testing I see some more errors from this PR. Will investigate.

a74nh committed:
Allow unpredicated movprfx with predicated instruction

a74nh committed:
Fix NI_Sve_ExtractVector

a74nh committed:
Fix for uses with zero masks (eg loads)

a74nh committed:
Fix tied register checking

JulieLeeMSFT committed:
Merge branch 'main' into movprfxvalidate_github

amanasifkhalid commented:
> However - when running the stress testing I see some more errors from this PR. Will investigate.

@a74nh I'm assuming you're getting a clean run now, right?

amanasifkhalid commented on src/coreclr/jit/emit.cpp:
```diff
@@ -7095,8 +7095,28 @@ unsigned emitter::emitEndCodeGen(Compiler*         comp,
 #endif
+#if defined(DEBUG) && defined(TARGET_ARM64)
+    instrDesc* prevId = nullptr;
+#endif // DEBUG && TARGET_ARM64
```
Nit:
```suggestion
#endif // defiined(DEBUG) && defined(TARGET_ARM64)
```

amanasifkhalid commented on src/coreclr/jit/emit.cpp:
```diff
@@ -7095,8 +7095,28 @@ unsigned emitter::emitEndCodeGen(Compiler*         comp,
+        {
+            assert(prevId->idIns() != INS_sve_movprfx);
+        }
+#endif // DEBUG && TARGET_ARM64
```
Nit:
```suggestion
#endif // defined(DEBUG) && defined(TARGET_ARM64)
```

amanasifkhalid commented on src/coreclr/jit/emitarm64sve.cpp:
```diff
@@ -18432,4 +18432,212 @@ void emitter::getInsSveExecutionCharacteristics(instrDesc* id, insExecutionChara
+#ifdef DEBUG
+/*****************************************************************************
+ *
+ *  Sanity check two instruction are valid when placed next to each other
```
Nit:
```suggestion
 *  Sanity check two instructions are valid when placed next to each other
```

amanasifkhalid commented on src/coreclr/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -2560,7 +2571,7 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 HWIntrinsicImmOpHelper helper(this, intrin.op3, node);
 
-                for (helper.EmitBegin(); !helper.Done(); helper.EmitCaseEnd())
+                for (helper.EmitBegin(); !helper.Done(); helper.EmitCaseEnd(), (targetReg != op1Reg) ? 2 : 1)
```
What is this syntax doing? It looks like the result of the ternary operator isn't being used.

a74nh replied:
That change should have been on the line above - the `HWIntrinsicImmOpHelper` constructor needs to take in the number of instructions that will be generated in each stage of the loop.

amanasifkhalid commented:
>I'll raise a bug and take a look at it - I have a reasonable idea of what it might be.
This PR shouldn't need to be blocked on it though.

Got it, thanks!

a74nh committed:
revert type of movprfx

a74nh committed:
fix comments

a74nh committed:
Fix use of immediate helpers

a74nh commented:
> I'll raise a bug and take a look at it - I have a reasonable idea of what it might be.

This was due to changes in #106125. I've fixed in this PR because it's related and simple enough - that PR changed the type of `movprfx` emitted when it shouldn't have. 

Also fixed up review comments.

Just checking now that all stress testing passes now.

amanasifkhalid reviewed (approved):
LGTM if stress tests pass. Thanks!

a74nh committed:
fix formatting

a74nh commented:
> LGTM if stress tests pass. Thanks!

Full stress testing for all APIs passing now.
ro: https://gist.github.com/a74nh/b690f44a2589873fce3066592d5141fa
and same for r too.
